src/storage/txn/scheduler.rs

```diff
@@ -309,33 +357,45 @@ impl<E: Engine> Scheduler<E> {
 
     /// Initiates an async operation to get a snapshot from the storage engine, then posts a
     /// `SnapshotFinished` message back to the event loop when it finishes.
-    fn get_snapshot(&mut self, cid: u64) {
-        let task = self.dequeue_task(cid);
+    fn get_snapshot(&self, cid: u64) {
+        let task = self.inner.dequeue_task(cid);
         let tag = task.tag;
         let ctx = task.context().clone();
-        let executor = self.fetch_executor(task.priority());
 
+        let executor = self.fetch_executor(task.priority());
+        let pool = executor.clone_pool();
+        let inner = self.clone();
         let cb = Box::new(move |(cb_ctx, snapshot)| {
             executor.execute(cb_ctx, snapshot, task);
         });
-        if let Err(e) = self.engine.async_snapshot(&ctx, cb) {
-            SCHED_STAGE_COUNTER_VEC
-                .with_label_values(&[tag, "async_snapshot_err"])
-                .inc();
 
-            error!("engine async_snapshot failed"; "err" => ?e);
-            self.finish_with_err(cid, e.into());
-        } else {
-            SCHED_STAGE_COUNTER_VEC
-                .with_label_values(&[tag, "snapshot"])
-                .inc();
-        }
+        pool.schedule(move |sched_ctx: &mut SchedContext<E>| {
+            if let Err(e) = sched_ctx.engine.async_snapshot(&ctx, cb) {
+                SCHED_STAGE_COUNTER_VEC
+                    .with_label_values(&[tag, "async_snapshot_err"])
+                    .inc();
+
+                error!("engine async_snapshot failed, err: {:?}", e);
+                notify_scheduler(
+                    inner,
+                    Msg::FinishedWithErr {
+                        cid,
+                        err: e.into(),
+                        tag,
+                    },
+                );
+            } else {
+                SCHED_STAGE_COUNTER_VEC
+                    .with_label_values(&[tag, "snapshot"])
+                    .inc();
+            }
+        });
     }
 
     /// Calls the callback with an error.
-    fn finish_with_err(&mut self, cid: u64, err: Error) {
-        debug!("command finished with error"; "cid" => cid);
-        let tctx = self.dequeue_task_context(cid);
+    fn finish_with_err(&self, cid: u64, err: Error) {
+        debug!("command cid={}, finished with error", cid);
+        let tctx = self.inner.dequeue_task_context(cid);
 
         SCHED_STAGE_COUNTER_VEC
```

Comment:
oh, I found another optimization place, we should use static metrics here @fredchenbj

Reply:
yes, and pr https://github.com/tikv/tikv/pull/3229 has tried before.